### PR TITLE
Solar rotation and sunpy coordinates - updated

### PR DESCRIPTION
**Pull request by wafels**
I copied the files that I actually touched in #2221 in to a new branch.  This PR updates the solar rotation code to use sunpy coordinates.

### CODE REVIEW

**Comment by wafels** [comment]
That should have read "updating the other differential rotation example".

File: examples/overplot_hek_polygon.py [hunk 1]
```diff
@@ -53,31 +55,22 @@
...
-rotated_boundary = np.zeros_like(ch_boundary)
-n = ch_boundary.shape[0]
-for i in range(0, n):
-    new_coords = rot_hpc(ch_boundary[i, 0] * u.arcsec,
-                         ch_boundary[i, 1] * u.arcsec,
-                         ch_date,
-                         aia_map.date)
-    rotated_boundary[i, 0] = new_coords[0].value
-    rotated_boundary[i, 1] = new_coords[1].value
+ch_boundary = SkyCoord([(eval(v[0]), eval(v[1]))*u.arcsec for v in p3], obstime=ch_date, frame=frames.Helioprojective)
```
Comment (Cadair) [inline examples/overplot_hek_polygon.py]: `eval`?! what is it?!

File: sunpy/physics/differential_rotation.py [hunk 2]
```diff
@@ -87,317 +78,73 @@ def diff_rot(duration, latitude, rot_type='howard', frame_time='sidereal'):
     if frame_time == 'synodic':
         rotation_deg -= 0.9856 * delta_days
 
-    #return Longitude((np.round(rotation_deg, 4)), u.deg)
-    return np.round(rotation_deg, 4) * u.deg
-
+    return rotation_deg * u.deg
```
Comment (Cadair) [inline sunpy/physics/differential_rotation.py]: does this not want to be a `Longitude` object?

File: sunpy/physics/differential_rotation.py [hunk 3]
```diff
@@ -87,317 +78,73 @@ def diff_rot(duration, latitude, rot_type='howard', frame_time='sidereal'):
...
-    earlier time.  Note that this function assumes that the data was observed
-    from the Earth or near Earth vicinity.  Specifically, data from SOHO and
-    STEREO observatories are not supported.  Note also that the function does
-    NOT use solar B0 and L0 values provided in source FITS files - these
-    quantities are calculated.
 
+def solar_rotate_coordinate(coordinate,
+                            new_observer_time,
+                            new_observer_location="earth", **diff_rot_kwargs):
+    """Given a coordinate on the Sun, calculate where that coordinate maps to
```
Comment (Cadair) [inline sunpy/physics/differential_rotation.py]: line break after `"""`.

File: sunpy/physics/differential_rotation.py [hunk 4]
```diff
@@ -87,317 +78,73 @@ def diff_rot(duration, latitude, rot_type='howard', frame_time='sidereal'):
...
+    at some later or earlier time, given the solar rotation profile.  Note that
+    if the new observer location is defined using a BaseCoordinateFrame or
```
Comment (Cadair) [inline sunpy/physics/differential_rotation.py]: I would link these classes, i.e:
```
`~astropy.coordinates.BaseCoordinateFrame`
`~astropy.coordinates.SkyCoord`
```

File: sunpy/physics/differential_rotation.py [hunk 5]
```diff
@@ -87,317 +78,73 @@ def diff_rot(duration, latitude, rot_type='howard', frame_time='sidereal'):
...
-    rot_type : {'howard' | 'snodgrass' | 'allen'}
-        | howard: Use values for small magnetic features from Howard et al.
-        | snodgrass: Use Values from Snodgrass et. al
-        | allen: Use values from Allen, Astrophysical Quantities, and simpler
-          equation.
-
-    frame_time : {'sidereal' | 'synodic'}
-        Choose type of day time reference frame.
-
+    coordinate : `~sunpy.coordinates`
```
Comment (Cadair) [inline sunpy/physics/differential_rotation.py]: type should just be `astropy.coordinates.SkyCoord` the description should say something like "any valid coordinate which is transformable to heliographic stonyhurst.

File: sunpy/physics/differential_rotation.py [hunk 6]
```diff
@@ -87,317 +78,73 @@ def diff_rot(duration, latitude, rot_type='howard', frame_time='sidereal'):
...
+        a sunpy coordinate
+    new_observer_time : `sunpy.time.time`
```
Comment (Cadair) [inline sunpy/physics/differential_rotation.py]: sunpy.time.time is not a thing. I normally just put 'a sunpy-compatible time'

File: sunpy/physics/differential_rotation.py [hunk 7]
```diff
@@ -87,317 +78,73 @@ def diff_rot(duration, latitude, rot_type='howard', frame_time='sidereal'):
...
+        date/time at which the input co-ordinate will be rotated to.
+    new_observer_location : None, str, BaseCoordinateFrame, SkyCoord
```
Comment (Cadair) [inline sunpy/physics/differential_rotation.py]: these should all be links to classes. also now you removed the handling in this function, None is no longer a valid option.

File: sunpy/physics/differential_rotation.py [hunk 8]
```diff
@@ -87,317 +78,73 @@ def diff_rot(duration, latitude, rot_type='howard', frame_time='sidereal'):
...
-    -----------
-    date : `sunpy.time.time`
-        the time at which to calculate the solar P, B0 angles and the
-        semi-diameter.
-
-    Returns
-    -------
-    A dictionary with the following keys with the following meanings:
+    # rotation.
+    if coordinate.obstime is None:
```
Comment (Cadair) [inline sunpy/physics/differential_rotation.py]: I wonder if you actually need to catch this here, the transformation will error when you try it?

**Comment by Cadair** [comment]
Looks great @wafels, I love how many lines of code this removes!

File: examples/overplot_hek_polygon.py (hunk 1 above)
Comment (wafels) [inline examples/overplot_hek_polygon.py]: changed to float

**Pull request merged**